### ISSUE TITLE
Suggestion: rename "pre/post-gradient" miners

### ISSUE THREAD

**Issue by zabzug-pfpt** [KevinMusgrave/pytorch-metric-learning#15]
Preface: I'm not at all engrained in the deep metric learning world, so if the terms "pre-gradient miner" and "post-gradient miner" are accepted and understood there, please disregard the rest of this comment!

I feel like the "pre-gradient" and "post-gradient" terms are confusing. The distinction seems to imply that the post-gradient miners actually _use_ the gradients to help select pairs/triplets. Furthermore, when people typically talk about gradients, it's usually in the context of backpropagating some loss (which, for pair/triplet training, can't be computed until the pairs/triplets have already been mined).

My understanding is that the pre/post-gradient distinction you are using is really focused on computational efficiency, and in particular, whether or not gradients are tracked when embeddings are calculated. In the case of pre-gradient miners, the answer is no (which means that embeddings need to be calculated again after the pairs/triplets have been mined). In the case of post-gradient miners, the answer is yes, and no extra computation is needed. 

That said, it seems like any of these miners _could_ be implemented in either a pre-gradient or post-gradient fashion, if you didn't care about computational efficiency. Additionally, it seems like any of them could be set up to accommodate differing (as in pre-gradient) or matched (as in post-gradient) input/output batch sizes.

Am I missing anything? I'm not sure I have a specific suggestion for different terminology right now, but I wanted to at least get a discussion started.

**Comment by KevinMusgrave** [comment]
Yeah you're basically right. Maybe names like these would be better?
```pre_gradient_miner``` --> ```batch_miner```
```post_gradient_miner``` --> ```tuple_miner```

**Comment by zabzug-pfpt** [comment]
Yeah, I think that terminology would be okay. Then you could explicitly specify that:
- a `batch_miner` will take a batch of `N` embeddings and return a subset `n` to use for pair/triplet mining. Without a `batch_miner`, `n == N`.
- a `tuple_miner` will take a batch of `n` embeddings and return `k` pairs/triplets to use for calculating the network loss. Without a `tuple_miner`, `k` will be all possible pairs/triplets (which I think would be `n(n-1)`?)

Are all of those statements correct?

**Comment by KevinMusgrave** [comment]
@zabzug-pfpt Yes, though right now the triplet-based losses don't collect all triplets by default. (You have to specify triplets_per_anchor = "all". Maybe I should make that the default instead.) Anyway that's beside the point. What you've written is clear, so when I make the change I'll put that in the documentation.

**Comment by KevinMusgrave** [comment]
@zabzug-pfpt The base classes are now BaseSubsetBatchMiner and BaseTupleMiner. For use with trainers, the keys in mining_funcs are "batch_miner" and "tuple_miner". (I should probably change "batch_miner" to "subset_batch_miner", so I'll do that in the next release.) 

I also updated the documentation and borrowed your definitions: https://kevinmusgrave.github.io/pytorch-metric-learning/miners/

Hopefully the names are clearer now!

**Comment by pavelgonchar** [comment]
Could you amend the examples accordingly? They still have the old miner names
`mining_funcs = {"post_gradient_miner": miner}`
https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/examples/example_MetricLossOnly.py#L69

**Comment by KevinMusgrave** [comment]
@pavelgonchar Thanks for pointing this out. I've updated the example scripts. This made me realize that it's easy to supply incorrect dictionary keys without consequence. So for the next release, I'm going to add assertions in the trainers to make sure dictionary keys have the correct form. I made an issue an for it: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/25#issue-584042486
